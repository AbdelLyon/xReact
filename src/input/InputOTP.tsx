import { Fragment } from "react/jsx-runtime";

import {
  InputOTPGroup,
  InputOTPSeparator,
  InputOTPSlot,
  InputOTP as InputOTPRoot,
} from "@/ui/input-otp";
import { cn } from "@/utils";

interface InputOTPProps {
  maxLength?: number;
  groupSize?: number;
  separator?: boolean;
  onChange?: (value: string) => void;
  onComplete?: (value: string) => void;
  defaultValue?: string;
  disabled?: boolean;
  label?: string;
  classNames?: {
    root?: string;
    label?: string;
    input?: string;
    separator?: string;
    group?: string;
    slot?: string;
  };
  [key: string]: any;
}

export function InputOTP({
  maxLength = 6,
  groupSize = 3,
  separator = true,
  onChange,
  onComplete,
  defaultValue = "",
  disabled = false,
  label,
  classNames = {},
  ...props
}: Readonly<InputOTPProps>) {
  const numberOfGroups = Math.ceil(maxLength / groupSize);

  const groups = Array.from({ length: numberOfGroups }, (_, groupIndex) => {
    const startIndex = groupIndex * groupSize;
    const groupSlots = Array.from({ length: groupSize }, (_, slotIndex) => {
      const index = startIndex + slotIndex;

      return index < maxLength ? (
        <InputOTPSlot key={index} className={classNames.slot} index={index} />
      ) : null;
    }).filter(Boolean);

    const group = (
      <InputOTPGroup key={`group-${groupIndex}`} className={classNames.group}>
        {groupSlots}
      </InputOTPGroup>
    );

    return (
      <Fragment key={`fragment-${groupIndex}`}>
        {group}
        {groupIndex < numberOfGroups - 1 && separator && (
          <InputOTPSeparator
            key={`sep-${groupIndex}`}
            className={classNames.separator}
          >
            {separator}
          </InputOTPSeparator>
        )}
      </Fragment>
    );
  });

  const handleChange = (value: string) => {
    onChange?.(value);
    if (value.length === maxLength) {
      onComplete?.(value);
    }
  };

  return (
    <div className={classNames.root}>
      {label && (
        <label className={cn("block mb-1", classNames.label)}>{label}</label>
      )}
      <InputOTPRoot
        className={classNames.input}
        defaultValue={defaultValue}
        disabled={disabled}
        maxLength={maxLength}
        onChange={handleChange}
        {...props}
      >
        {groups}
      </InputOTPRoot>
    </div>
  );
}
