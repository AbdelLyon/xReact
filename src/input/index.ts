import "@/index.css";
export { Input, Checkbox, Switch, Textarea } from "@/input/Input";
export { InputOTP } from "@/input/InputOTP";
export { Radio, type RadioOptions } from "@/input/Radio";
