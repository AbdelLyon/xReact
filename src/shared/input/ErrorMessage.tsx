import { cn } from "@/utils";

export const ErrorMessage = ({
  error,
  classNameError,
}: {
  error: string;
  classNameError?: string;
}) => (
  <p className={cn("text-error mt-1 text-start", classNameError)}>{error}</p>
);
