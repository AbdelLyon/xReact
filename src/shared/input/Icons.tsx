import { Eye, EyeOff, Mail, Search } from "lucide-react";
import { Dispatch } from "react";

import { cn } from "@/utils";
import { InputType } from "@/types/formTypes";

type PropsIcon = { className?: string; size?: number };

export const PassowrdIcon = ({
  className,
  setIsPasswordVisible,
  isPasswordVisible,
  size = 16,
}: PropsIcon & {
  setIsPasswordVisible: Dispatch<React.SetStateAction<boolean>>;
  isPasswordVisible: boolean;
}) => {
  const togglePasswordVisibility = () => {
    setIsPasswordVisible((prev) => !prev);
  };

  return (
    <>
      {isPasswordVisible ? (
        <EyeOff
          className={cn("text-muted-foreground right-3 absolute", className)}
          id="eye-icon"
          size={size}
          style={{ top: "50%", transform: "translateY(-50%)" }}
          onClick={togglePasswordVisibility}
        />
      ) : (
        <Eye
          className={cn("text-muted-foreground right-3 absolute", className)}
          size={size}
          style={{ top: "50%", transform: "translateY(-50%)" }}
          onClick={togglePasswordVisibility}
        />
      )}
    </>
  );
};

const SearchIcon = ({ className, size = 13 }: PropsIcon) => (
  <Search
    className={cn("text-muted-foreground right-3 absolute", className)}
    size={size}
    style={{ top: "50%", transform: "translateY(-50%)" }}
  />
);

const MailIcon = ({ className, size = 13 }: PropsIcon) => (
  <Mail
    className={cn("text-muted-foreground right-3 absolute", className)}
    size={size}
    style={{ top: "50%", transform: "translateY(-50%)" }}
  />
);

type Props = {
  type: InputType;
  isPasswordVisible: boolean;
  setIsPasswordVisible: any;
  error?: string;
  classNameIcon?: string;
  value?: string | number;
  iconSise?: number;
};

export const InputIcon = ({
  type,
  isPasswordVisible,
  setIsPasswordVisible,
  error,
  classNameIcon,
  value,
  iconSise,
}: Props) => {
  return (
    <>
      {type === "password" && (
        <div>
          <PassowrdIcon
            className={cn(
              { "text-error": error && error !== "" },
              classNameIcon,
            )}
            isPasswordVisible={isPasswordVisible}
            setIsPasswordVisible={setIsPasswordVisible}
            size={iconSise}
          />
        </div>
      )}
      {type === "search" && value === "" && (
        <SearchIcon
          className={cn({ "text-error": error && error !== "" }, classNameIcon)}
          size={iconSise}
        />
      )}
      {type === "email" && (
        <MailIcon
          className={cn({ "text-error": error && error !== "" }, classNameIcon)}
          size={iconSise}
        />
      )}
    </>
  );
};
