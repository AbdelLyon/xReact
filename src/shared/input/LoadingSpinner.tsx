import { Loader2 } from "lucide-react";

export const LoadingSpinner = () => (
  <div className="absolute bottom-0 left-0 top-1 flex size-full items-center justify-center backdrop-blur-[1px]">
    <div className="absolute bottom-2 right-0 z-10 flex w-full justify-center text-center text-primary">
      <Loader2 className="animate-spin font-extrabold" />
    </div>
  </div>
);
