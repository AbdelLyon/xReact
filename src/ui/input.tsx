import * as React from "react";

import { cn } from "@/utils";

export interface InputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {}

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  ({ className, type, ...props }, ref) => {
    return (
      <input
        ref={ref}
        className={cn(
          "flex h-11 w-full rounded-md border border-border bg-white dark:bg-background focus:border-outline px-3 py-2 text-[13px] ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50",

          className,
        )}
        type={type}
        {...props}
      />
    );
  },
);

Input.displayName = "Input";

export { Input };
