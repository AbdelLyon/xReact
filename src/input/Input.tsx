import { HTMLAttributes, forwardRef, useEffect, useState } from "react";
import { passwordStrength } from "check-password-strength";

import { ErrorMessage } from "../shared/input/ErrorMessage";
import { InputIcon } from "../shared/input/Icons";

import { cn } from "@/utils";
import { Input as InputRoot } from "@/ui/input";
import { CheckboxPrimitive, Checkbox as CheckboxRoot } from "@/ui/checkbox";
import { SwitchPrimitives, SwitchRoot } from "@/ui/switch";
import { TextareaRoot } from "@/ui/textarea";
import { PasswordStrength } from "@/shared/input/PasswordStrength";
import {
  CheckboxAndSwitchProps,
  InputProps,
  TextareaProps,
} from "@/types/formTypes";

const Input = forwardRef<
  HTMLInputElement,
  InputProps & HTMLAttributes<HTMLInputElement>
>(
  (
    {
      type = "text",
      classNames,
      onChange,
      iconSize,
      error,
      width = "100%",
      isPasswordStrength,
      min,
      max,
      ...props
    },
    ref,
  ) => {
    const [isPasswordVisible, setIsPasswordVisible] = useState(false);
    const [strength, setStrength] = useState(0);
    const [inputValue, setInputValue] = useState("");

    useEffect(() => {
      if (typeof ref === "object" && ref?.current)
        setInputValue(ref.current.value);

      if (type === "password" && typeof inputValue === "string") {
        const passwordStrengthId = passwordStrength(inputValue).id;

        setStrength(passwordStrengthId);
      }
    }, [inputValue, type, ref]);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(e.target.value);
      onChange?.(e);
    };

    return (
      <div className={classNames?.root}>
        {props.label && type !== "checkbox" && (
          <label className={classNames?.label}>{props.label}</label>
        )}
        <div
          className={cn("relative", {
            "flex items-center gap-1": type === "checkbox",
          })}
          style={{ width }}
        >
          <InputRoot
            ref={ref}
            className={cn(
              "text-sm",
              { "border-error": error },
              {
                "size-4 rounded-sm cursor-pointer": type === "checkbox",
              },
              classNames?.input,
            )}
            placeholder={props.placeholder}
            type={isPasswordVisible ? "text" : type}
            value={props.value}
            onChange={handleChange}
            {...props}
            max={max}
            min={min}
          />

          <InputIcon
            classNameIcon={classNames?.icon}
            error={error}
            iconSise={iconSize}
            isPasswordVisible={isPasswordVisible}
            setIsPasswordVisible={setIsPasswordVisible}
            type={type}
            value={props.value}
          />
          {props.label && type === "checkbox" && (
            <label className={classNames?.label}>{props.label}</label>
          )}
        </div>
        {isPasswordStrength && (
          <PasswordStrength className="my-1 px-0.5" strength={strength} />
        )}

        {error && type !== "checkbox" && <ErrorMessage error={error} />}
      </div>
    );
  },
);

Input.displayName = "Input";

const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(
  ({ className, ...props }, ref) => {
    return <TextareaRoot ref={ref} {...props} className={className} />;
  },
);

Textarea.displayName = "Textarea";

const Switch = forwardRef<
  React.ElementRef<typeof SwitchPrimitives.Root>,
  CheckboxAndSwitchProps
>(({ className, ...props }, ref) => {
  return <SwitchRoot ref={ref} {...props} className={className} />;
});

Switch.displayName = "Switch";

const Checkbox = forwardRef<
  React.ElementRef<typeof CheckboxPrimitive.Root>,
  CheckboxAndSwitchProps
>(({ className, ...props }, ref) => {
  return <CheckboxRoot ref={ref} {...props} className={className} />;
});

Checkbox.displayName = "Checkbox";

export { Input, Checkbox, Switch, Textarea };
