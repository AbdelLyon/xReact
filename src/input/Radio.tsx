import { RadioGroup, RadioGroupItem } from "@/ui/radio-group";

export type RadioOptions = { value: string; label: string };

interface GenericRadioGroupProps {
  options: RadioOptions[];
  defaultValue?: string;
  name: string;
  onChange: (value: string) => void;
  className?: string;
}

export function Radio({
  options,
  defaultValue,
  name,
  onChange,
  className = "",
}: Readonly<GenericRadioGroupProps>) {
  return (
    <RadioGroup
      className={className}
      defaultValue={defaultValue}
      name={name}
      onValueChange={onChange}
    >
      {options.map((option, index) => (
        <div
          key={option.value}
          className={`flex w-max items-center space-x-2 ${options.length === 1 ? "justify-start" : "justify-between"}`}
        >
          <RadioGroupItem id={`${name}-${index}`} value={option.value} />
          <label htmlFor={`${name}-${index}`}>{option.label}</label>
        </div>
      ))}
    </RadioGroup>
  );
}
