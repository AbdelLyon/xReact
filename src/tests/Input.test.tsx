import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import { Input } from "@/input";

describe("Input", () => {
  it("should match the snapshot with default props", () => {
    const { asFragment } = render(
      <Input placeholder="Enter text" onChange={() => {}} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with a password field", () => {
    const { asFragment } = render(
      <Input placeholder="Password" type="password" onChange={() => {}} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with an icon", () => {
    const { asFragment } = render(
      <Input
        className="rounded border"
        iconSize={24}
        placeholder="Search"
        type="text"
        onChange={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it("should match the snapshot with error message", () => {
    const { asFragment } = render(
      <Input
        error="This field is required"
        placeholder="Enter text"
        type="text"
        onChange={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with password strength indicator", () => {
    const { asFragment } = render(
      <Input
        isPasswordStrength
        placeholder="Password"
        type="password"
        onChange={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with checkbox type", () => {
    const { asFragment } = render(
      <Input
        className="ml-2"
        label="I agree to the terms"
        type="checkbox"
        onChange={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
