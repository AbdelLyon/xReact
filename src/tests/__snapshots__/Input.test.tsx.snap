// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should match the snapshot with a password field 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full rounded-md border border-border bg-white dark:bg-background focus:border-outline px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 text-sm"
        placeholder="Password"
        type="password"
        value=""
      />
      <div>
        <svg
          class="lucide lucide-eye text-muted-foreground right-3 absolute"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="top: 50%; transform: translateY(-50%);"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should match the snapshot with an icon 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full border-border bg-white dark:bg-background focus:border-outline px-3 py-2 text-[13px] ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded border"
        placeholder="Search"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should match the snapshot with checkbox type 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative flex items-center gap-1"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full rounded-md border border-border bg-white dark:bg-background focus:border-outline px-3 py-2 text-[13px] ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 ml-2"
        label="I agree to the terms"
        type="checkbox"
        value=""
      />
      <label>
        I agree to the terms
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full rounded-md border border-border bg-white dark:bg-background focus:border-outline px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 text-sm"
        placeholder="Enter text"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should match the snapshot with error message 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full rounded-md border bg-white dark:bg-background focus:border-outline px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 text-sm border-error"
        placeholder="Enter text"
        type="text"
        value=""
      />
    </div>
    <p
      class="text-error mt-1 text-start"
    >
      This field is required
    </p>
  </div>
</DocumentFragment>
`;

exports[`Input > should match the snapshot with password strength indicator 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative"
      style="width: 100%;"
    >
      <input
        class="flex h-11 w-full rounded-md border border-border bg-white dark:bg-background focus:border-outline px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:opacity-50 placeholder:text-[13px] focus-visible:outline-none focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 text-sm"
        placeholder="Password"
        type="password"
        value=""
      />
      <div>
        <svg
          class="lucide lucide-eye text-muted-foreground right-3 absolute"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="top: 50%; transform: translateY(-50%);"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
      </div>
    </div>
    <div
      class="flex space-x-1 my-1 px-0.5"
    >
      <div
        class="h-2 flex-1 transition-all duration-300 rounded-sm bg-error"
      />
      <div
        class="h-2 flex-1 transition-all duration-300 rounded-sm bg-light dark:bg-card"
      />
      <div
        class="h-2 flex-1 transition-all duration-300 rounded-sm bg-light dark:bg-card"
      />
      <div
        class="h-2 flex-1 transition-all duration-300 rounded-sm bg-light dark:bg-card"
      />
    </div>
  </div>
</DocumentFragment>
`;
